problem/0000-metrics-v1.md

```diff
@@ -0,0 +1,136 @@
+  - Peer Identity
+  - Peer K8s Labels? How?
+- Close Count, by the above as well as an errno.
+- Connect Latency Histogram, by the above
+- Connection Age Histogram, by the above
+- Bytes Transferred, by the above
+
+#### HTTP
+
+- Request count, by
+  - Interface: `inbound | outbound`
+  - Authority
+  - Workload
+  - Service
+  - Service profile route
+  - Peer Identity
+  - Peer K8s Labels? How?
+- Response count, by
+  - The above, plus
+  - Status Code; [100, 599]
+  - Classification: success | failure
+- Response Latency Histogram
+  - Should users be able to query at arbitrary percentiles?
+
+##### gRPC
+
+How do we represent the gRPC-specific metadata that Linkerd exposes? It seems like it should be
+distinct from (but in addition to) the HTTP-level abstractions? Or should HTTP metrics have a
+hole to punch through arbitrary metadata?
+
+Should we provide gRPC stats scoped by gRPC service name? We can infer these via known path
+structure when the appropriate headers are present.
+
+#### K8s services
+
+One of the main complexities in our metrics comes from handling Service-specific metadata.
+Currently all service-specific metadata is measured on the client-side, even when the querier
+asks about server-side resources. It is effectively impossible to query for server-side metrics
+for a service, since we allow pods to exist in more than one service. This is
+[fixable][rfc-src-metrics], however, if servers are able to discover metadata for inbound
+traffic.
+
+It seems like we should not support querying service metrics as a resource scope, though, because
+we cannot measure traffic _on_ a Service. We can measure outbound traffic from a set of client
+pods to a Service; and we can measure inbound traffic to a set of server pods that is targetted
+at a Service. I think our API should be explicit about these constraints.
+
+[rfc-src-metrics]: https://github.com/linkerd/rfc/pull/15
+
+### Prior Art
+
+#### Linkerd Stat
+
+Linkerd's [`Stat`][stat] API is resource oriented, but it only exposes an _outbound_ side, so
+there's no way to
+
+Why is `skip_stats` part of the API?
+
+The `tcp_stats` flag is a legacy of the CLI-specific nature of the command... it seems like we
+should either support a list of protocols or have protocol-specific endpoints.
+
+[l5d-stat]: https://github.com/linkerd/linkerd2/blob/dacf87e084a55757aab9cfa8556053495e43e207/proto/public.proto#L348-L427
+
+#### SMI
+
+[SMI's metrics][smi-spec] appear to be stringly typed, in that their are a few well-defined
+(http-specific?) metrics, but the base set of metrics is insufficient to satisfy Linkerd's CLI
+needs.
+
+However, the way it's resource-oriented is very appealing.
+
+It does not yet expose any time-series functionality.
+
+[smi-spec]: https://github.com/servicemeshinterface/smi-spec/blob/d5dd526a1a19c784065dc3fbac05a52decc01735/traffic-metrics.md
+
+#### Prometheus
+
+Can we expose a [Prometheus API][prom-api] scoped under each resource? E.g.
```

Comment:
I'm wondering if we can get everything we want from recording rules. That provides the stable API and solves some other issues. We'd still need to parse and such, but the surface area would be shrunk.

Reply:
Recording rules are probably a good way to try this, yeah.

But, I think there are opportunities for us to be a bit clever about our storage/query patterns to take advantage of our understanding of the data model.

For example: Can we reduce the number of time-series that have to be stored/in memory by _only_ storing the pod name in a metric label? What if, instead, we wrote a single `pod{name="",parent_kind="...",parent_name="...",label_foo="...",label_bar="...} 1`. If you want to do a deploy-specific lookup, instead of looking up a `request_total{deployment="foo"}`, we would first expand the list of pods for the deployment and look up by `{pod=".."}`...

This is, of course, totally hypothetical and we'd want to dig on access patterns before optimizing this way. But my point is this plan gives us enough room to get creative...